File: src/strands/vended_plugins/context_offloader/storage.py
"""Storage backends for offloaded tool result content.

This module defines the Storage protocol and provides three built-in
implementations: file-based, in-memory, and S3 storage. Each content block
from a tool result is stored individually with its content type preserved.

Example:
    ```python
    from strands.vended_plugins.context_offloader import (
        FileStorage,
        InMemoryStorage,
        S3Storage,
    )

    # File-based storage
    storage = FileStorage(artifact_dir="./artifacts")
    ref = storage.store("tool_123_0", b"large output content...", "text/plain")
    content, content_type = storage.retrieve(ref)

    # In-memory storage (useful for testing and serverless)
    storage = InMemoryStorage()

    # S3 storage
    storage = S3Storage(bucket="my-bucket", prefix="artifacts/")
    ```
"""

import json
import re
import threading
import time
from pathlib import Path
from typing import Any, Protocol, runtime_checkable

import boto3
from botocore.config import Config as BotocoreConfig
from botocore.exceptions import ClientError


def _sanitize_id(raw_id: str) -> str:
    """Sanitize an ID for safe use in filenames and object keys.

    Replaces path separators, parent directory references, and other
    unsafe characters with underscores.

    Args:
        raw_id: The raw ID string.

    Returns:
        A sanitized string safe for use in filenames.
    """
    sanitized = raw_id.replace("..", "_").replace("/", "_").replace("\\", "_")
    sanitized = re.sub(r"[^\w\-.]", "_", sanitized)
    return sanitized


@runtime_checkable
class Storage(Protocol):
    """Backend for storing and retrieving offloaded content blocks.

    Each content block from a tool result is stored individually with its
    content type preserved. The SDK ships three built-in implementations:
    ``InMemoryStorage``, ``FileStorage``, and ``S3Storage``. Implement this
    protocol to create custom storage backends (e.g., Redis, DynamoDB).

    Lifecycle:
        This protocol intentionally does not include eviction or deletion methods.
        Stored content accumulates for the lifetime of the storage instance. For
        long-running agents, create a new storage instance per session or use a
        backend with built-in lifecycle management (e.g., S3 lifecycle policies).
    """

    def store(self, key: str, content: bytes, content_type: str = "text/plain") -> str:
        """Store content and return a reference identifier.

        Args:
            key: A unique key for this content block.
            content: The raw content bytes to store.
            content_type: MIME type of the content (e.g., "text/plain",
                "application/json", "image/png", "application/pdf").

        Returns:
            A reference string that can be used to retrieve the content later.
        """
        ...

    def retrieve(self, reference: str) -> tuple[bytes, str]:
        """Retrieve stored content by reference.

        Args:
            reference: The reference returned by a previous store() call.

        Returns:
            A tuple of (content bytes, content type).

        Raises:
            KeyError: If the reference is not found.
        """
        ...


class FileStorage:
    """Store offloaded content as files on disk.

    Files are written to the configured artifact directory with unique names.
    File extensions are derived from the content type. A ``.metadata.json``
    sidecar file tracks content types so they survive process restarts.

    Args:
        artifact_dir: Directory path where artifact files will be stored.
    """

    _METADATA_FILE = ".metadata.json"

    def __init__(self, artifact_dir: str = "./artifacts") -> None:
        """Initialize file-based storage.

        Args:
            artifact_dir: Directory path where artifact files will be stored.
        """
        self._artifact_dir = Path(artifact_dir)
        self._counter: int = 0
        self._lock = threading.Lock()
        self._content_types: dict[str, str] = self._load_metadata()

    @staticmethod
    def _extension_for(content_type: str) -> str:
        """Return a file extension for the given content type."""
        if content_type == "text/plain":
            return ".txt"
        return f".{content_type.split('/')[-1]}"

    def store(self, key: str, content: bytes, content_type: str = "text/plain") -> str:
        """Store content as a file and return the filename as reference.

        Args:
            key: A unique key for this content block.
            content: The raw content bytes to store.
            content_type: MIME type of the content.

        Returns:
            The filename (not full path) used as the reference.
        """
        self._artifact_dir.mkdir(parents=True, exist_ok=True)

        sanitized_key = _sanitize_id(key)
        timestamp_ms = int(time.time() * 1000)
        ext = self._extension_for(content_type)
        with self._lock:
            self._counter += 1
            counter = self._counter
            filename = f"{timestamp_ms}_{counter}_{sanitized_key}{ext}"
            self._content_types[filename] = content_type
            self._save_metadata()

        file_path = self._artifact_dir / filename
        file_path.write_bytes(content)

        return filename

    def retrieve(self, reference: str) -> tuple[bytes, str]:
        """Retrieve content from a stored file.

        Args:
            reference: The filename reference returned by store().

        Returns:
            A tuple of (content bytes, content type).

        Raises:
            KeyError: If the file does not exist.
        """
        file_path = (self._artifact_dir / reference).resolve()
        if not file_path.is_relative_to(self._artifact_dir.resolve()):
            raise KeyError(f"Reference not found: {reference}")
        if not file_path.is_file():
            raise KeyError(f"Reference not found: {reference}")
        content_type = self._content_types.get(reference, "application/octet-stream")
        return file_path.read_bytes(), content_type

    def _load_metadata(self) -> dict[str, str]:
        """Load content type metadata from the sidecar file."""
        metadata_path = self._artifact_dir / self._METADATA_FILE
        if metadata_path.is_file():
            try:
                result: dict[str, str] = json.loads(metadata_path.read_text(encoding="utf-8"))
                return result
            except (json.JSONDecodeError, OSError):
                return {}
        return {}

    def _save_metadata(self) -> None:
        """Save content type metadata to the sidecar file."""
        metadata_path = self._artifact_dir / self._METADATA_FILE
        metadata_path.write_text(json.dumps(self._content_types), encoding="utf-8")


class InMemoryStorage:
    """Store offloaded content in memory.

    Useful for testing and serverless environments where disk access
    is not available or not desired. Thread-safe.

    Note:
        Content accumulates for the lifetime of this instance. For long-running
        agents, consider creating a new instance per session or switching to
        ``FileStorage`` or ``S3Storage`` for persistent storage with external
        lifecycle management.
    """

    def __init__(self) -> None:
        """Initialize in-memory storage."""
        self._store: dict[str, tuple[bytes, str]] = {}
        self._counter: int = 0
        self._lock = threading.Lock()

    def store(self, key: str, content: bytes, content_type: str = "text/plain") -> str:
        """Store content in memory and return a reference.

        Args:
            key: A unique key for this content block.
            content: The raw content bytes to store.
            content_type: MIME type of the content.

        Returns:
            A unique reference string.
        """
        with self._lock:
            self._counter += 1
            reference = f"mem_{self._counter}_{key}"
            self._store[reference] = (content, content_type)
        return reference

    def retrieve(self, reference: str) -> tuple[bytes, str]:
        """Retrieve content from memory.

        Args:
            reference: The reference returned by store().

        Returns:
            A tuple of (content bytes, content type).

        Raises:
            KeyError: If the reference is not found.
        """
        with self._lock:
            if reference not in self._store:
                raise KeyError(f"Reference not found: {reference}")
            return self._store[reference]

    def clear(self) -> None:
        """Remove all stored content.

        Call this to free memory when offloaded results are no longer needed,
        e.g., between sessions or after an invocation completes.
        """
        with self._lock:
            self._store.clear()


class S3Storage:
    """Store offloaded content in Amazon S3.

    Objects are stored with unique keys under the configured prefix.
    Content type is preserved as S3 object metadata.

    Args:
        bucket: S3 bucket name.
        prefix: S3 key prefix for organizing stored artifacts.
        boto_session: Optional boto3 session. If not provided, a new session
            is created using the given region_name.
        boto_client_config: Optional botocore client configuration.
        region_name: AWS region. Used only when boto_session is not provided.

    Example:
        ```python
        from strands.vended_plugins.context_offloader import S3Storage

        storage = S3Storage(
            bucket="my-agent-artifacts",
            prefix="tool-results/",
        )
        ```
    """

    def __init__(
        self,
        bucket: str,
        prefix: str = "",
        boto_session: boto3.Session | None = None,
        boto_client_config: BotocoreConfig | None = None,
        region_name: str | None = None,
    ) -> None:
        """Initialize S3-based storage.

        Args:
            bucket: S3 bucket name.
            prefix: S3 key prefix for organizing stored artifacts.
            boto_session: Optional boto3 session. If not provided, a new session
                is created using the given region_name.
            boto_client_config: Optional botocore client configuration.
            region_name: AWS region. Used only when boto_session is not provided.
        """
        self._bucket = bucket
        self._prefix = prefix.strip("/")
        if self._prefix:
            self._prefix += "/"

        session = boto_session or boto3.Session(region_name=region_name)

        if boto_client_config:
            existing_user_agent = getattr(boto_client_config, "user_agent_extra", None)
            new_user_agent = f"{existing_user_agent} strands-agents" if existing_user_agent else "strands-agents"
            client_config = boto_client_config.merge(BotocoreConfig(user_agent_extra=new_user_agent))
        else:
            client_config = BotocoreConfig(user_agent_extra="strands-agents")

        self._client: Any = session.client(service_name="s3", config=client_config)
        self._counter: int = 0
        self._lock = threading.Lock()

    def store(self, key: str, content: bytes, content_type: str = "text/plain") -> str:
        """Store content as an S3 object and return the object key as reference.

        Args:
            key: A unique key for this content block.
            content: The raw content bytes to store.
            content_type: MIME type of the content.

        Returns:
            The S3 object key used as the reference.

        Raises:
            botocore.exceptions.ClientError: If the S3 operation fails (e.g., bucket
                does not exist, permission denied).
        """
        sanitized_key = _sanitize_id(key)
        timestamp_ms = int(time.time() * 1000)
        with self._lock:
            self._counter += 1
            counter = self._counter
        s3_key = f"{self._prefix}{timestamp_ms}_{counter}_{sanitized_key}"

        self._client.put_object(
            Bucket=self._bucket,
            Key=s3_key,
            Body=content,
            ContentType=content_type,
        )

        return s3_key

    def retrieve(self, reference: str) -> tuple[bytes, str]:
        """Retrieve content from an S3 object.

        Args:
            reference: The S3 object key returned by store().

        Returns:
            A tuple of (content bytes, content type).

        Raises:
            KeyError: If the object does not exist.
        """
        try:
            response = self._client.get_object(Bucket=self._bucket, Key=reference)
            content: bytes = response["Body"].read()
            content_type: str = response.get("ContentType", "application/octet-stream")
            return content, content_type
        except ClientError as e:
            if e.response["Error"]["Code"] == "NoSuchKey":
                raise KeyError(f"Reference not found: {reference}") from e
            raise

File: strands-ts/src/vended-plugins/context-offloader/storage.ts
/**
 * Storage backends for offloaded tool result content.
 *
 * This module defines the {@link Storage} interface and provides three built-in
 * implementations: {@link InMemoryStorage}, {@link FileStorage}, and {@link S3Storage}.
 * Each content block from a tool result is stored individually with its content type preserved.
 */

import type { Sandbox } from '../../sandbox/base.js'

/**
 * Backend for storing and retrieving offloaded content blocks.
 *
 * Implement this interface to create custom storage backends (e.g., Redis, DynamoDB).
 * The SDK ships three built-in implementations: {@link InMemoryStorage},
 * {@link FileStorage}, and {@link S3Storage}.
 */
export interface Storage {
  /**
   * Store content and return a reference identifier.
   *
   * @param key - Unique key for this content block
   * @param content - Raw content bytes to store
   * @param contentType - MIME type of the content (e.g., "text/plain", "image/png")
   * @returns Reference string for later retrieval
   */
  store(key: string, content: Uint8Array, contentType?: string): Promise<string>

  /**
   * Retrieve previously stored content by reference.
   *
   * @param reference - Reference returned by a previous {@link store} call
   * @returns Content bytes and content type
   * @throws Error if the reference is not found
   */
  retrieve(reference: string): Promise<{ content: Uint8Array; contentType: string }>
}

function sanitizeId(rawId: string): string {
  return rawId
    .replace(/\.\./g, '_')
    .replace(/[/\\]/g, '_')
    .replace(/[^\w\-.]/g, '_')
}

/**
 * In-memory storage backend.
 *
 * Useful for testing and serverless environments where disk access is not available.
 * Content accumulates for the lifetime of this instance; call {@link clear} to free memory.
 */
export class InMemoryStorage implements Storage {
  private _store = new Map<string, { content: Uint8Array; contentType: string }>()
  private _counter = 0

  /** {@inheritdoc} */
  async store(key: string, content: Uint8Array, contentType: string = 'text/plain'): Promise<string> {
    this._counter++
    const reference = `mem_${this._counter}_${key}`
    this._store.set(reference, { content, contentType })
    return reference
  }

  /** {@inheritdoc} */
  async retrieve(reference: string): Promise<{ content: Uint8Array; contentType: string }> {
    const entry = this._store.get(reference)
    if (!entry) {
      throw new Error(`Reference not found: ${reference}`)
    }
    return entry
  }

  /** Remove all stored content. */
  clear(): void {
    this._store.clear()
  }
}

/**
 * File-based storage backend.
 *
 * Stores offloaded content as files on the host filesystem, or through a configured
 * {@link Sandbox}. File extensions are derived from the content type. A `.metadata.json`
 * sidecar file tracks content types across restarts. References are file paths preserving
 * the configured artifact directory form.
 *
 * When used by {@link ContextOffloader} without an explicit sandbox, FileStorage is
 * bound to each agent's sandbox, which may be the default NotASandboxLocalEnvironment.
 */
export interface FileStorageOptions {
  /** Directory path where artifact files will be stored. Defaults to `./artifacts`. */
  artifactDir?: string
  /** Sandbox to use for direct store/retrieve calls. */
  sandbox?: Sandbox
}

export class FileStorage implements Storage {
  private static readonly METADATA_FILE = '.metadata.json'
  private readonly _artifactDir: string
  private readonly _sandbox: Sandbox | undefined
  private _counter = 0
  private _contentTypes: Record<string, string> = {}
  private _metadataLoaded = false
  private _metadataWriteChain: Promise<void> = Promise.resolve()

  constructor(options: string | FileStorageOptions = './artifacts') {
    if (typeof options === 'string') {
      this._artifactDir = options
      this._sandbox = undefined
    } else {
      this._artifactDir = options.artifactDir ?? './artifacts'
      this._sandbox = options.sandbox
    }
  }

  /**
   * Return a storage instance bound to the provided sandbox.
   *
   * Instances constructed with an explicit sandbox keep using that sandbox. Detached
   * instances produce a new storage object so a shared ContextOffloader can isolate
   * artifacts for each agent sandbox.
   *
   * @param sandbox - Sandbox to use for the returned storage instance.
   */
  forSandbox(sandbox: Sandbox): FileStorage {
    if (this._sandbox) return this
    return new FileStorage({ artifactDir: this._artifactDir, sandbox })
  }

  private static _extensionFor(contentType: string): string {
    if (contentType === 'text/plain') return '.txt'
    return `.${contentType.split('/').pop()}`
  }

  private _artifactPath(filename: string): string {
    return `${this._artifactDir.replace(/\/+$/, '')}/${filename}`
  }

  private async _ensureDir(): Promise<typeof import('node:fs/promises')> {
    const fs = await import('node:fs/promises')
    await fs.mkdir(this._artifactDir, { recursive: true })
    if (!this._metadataLoaded) {
      this._contentTypes = await this._loadMetadata(fs)
      this._metadataLoaded = true
    }
    return fs
  }

  private async _ensureSandbox(): Promise<Sandbox> {
    const sandbox = this._sandbox
    if (!sandbox) {
      throw new Error('FileStorage requires a Sandbox to be configured on the storage instance.')
    }
    if (!this._metadataLoaded) {
      this._contentTypes = await this._loadSandboxMetadata(sandbox)
      this._metadataLoaded = true
    }
    return sandbox
  }

  private async _loadMetadata(fs: typeof import('node:fs/promises')): Promise<Record<string, string>> {
    const path = await import('node:path')
    const metadataPath = path.join(this._artifactDir, FileStorage.METADATA_FILE)
    try {
      const raw = await fs.readFile(metadataPath, 'utf-8')
      return JSON.parse(raw) as Record<string, string>
    } catch {
      return {}
    }
  }

  private async _loadSandboxMetadata(sandbox: Sandbox): Promise<Record<string, string>> {
    try {
      const raw = await sandbox.readText(this._artifactPath(FileStorage.METADATA_FILE))
      return JSON.parse(raw) as Record<string, string>
    } catch {
      return {}
    }
  }

  private async _saveMetadata(fs: typeof import('node:fs/promises')): Promise<void> {
    const path = await import('node:path')
    const metadataPath = path.join(this._artifactDir, FileStorage.METADATA_FILE)
    await fs.writeFile(metadataPath, JSON.stringify(this._contentTypes), 'utf-8')
  }

  private async _saveSandboxMetadata(sandbox: Sandbox): Promise<void> {
    await sandbox.writeText(this._artifactPath(FileStorage.METADATA_FILE), JSON.stringify(this._contentTypes))
  }

  /** {@inheritdoc} */
  async store(key: string, content: Uint8Array, contentType: string = 'text/plain'): Promise<string> {
    if (this._sandbox) {
      const sandbox = await this._ensureSandbox()
      const filename = `${Date.now()}_${++this._counter}_${sanitizeId(key)}${FileStorage._extensionFor(contentType)}`
      const filePath = this._artifactPath(filename)

      this._contentTypes[filename] = contentType
      this._metadataWriteChain = this._metadataWriteChain.then(() => this._saveSandboxMetadata(sandbox))
      await this._metadataWriteChain

      await sandbox.writeFile(filePath, content)

      return filePath
    }

    const fs = await this._ensureDir()
    const path = await import('node:path')

    const sanitizedKey = sanitizeId(key)
    const timestampMs = Date.now()
    this._counter++
    const ext = FileStorage._extensionFor(contentType)
    const filename = `${timestampMs}_${this._counter}_${sanitizedKey}${ext}`

    this._contentTypes[filename] = contentType
    this._metadataWriteChain = this._metadataWriteChain.then(() => this._saveMetadata(fs))
    await this._metadataWriteChain

    const filePath = path.join(this._artifactDir, filename)
    await fs.writeFile(filePath, content)

    return filePath
  }

  /** {@inheritdoc} */
  async retrieve(reference: string): Promise<{ content: Uint8Array; contentType: string }> {
    if (this._sandbox) {
      const sandbox = await this._ensureSandbox()

      if (!reference.startsWith(`${this._artifactDir.replace(/\/+$/, '')}/`) || reference.includes('..')) {
        throw new Error(`Reference not found: ${reference}`)
      }

      const filename = reference.split('/').pop()!

      try {
        const content = await sandbox.readFile(reference)
        const contentType = this._contentTypes[filename] ?? 'application/octet-stream'
        return { content, contentType }
      } catch {
        throw new Error(`Reference not found: ${reference}`)
      }
    }

    const fs = await this._ensureDir()
    const path = await import('node:path')

    const filePath = path.resolve(this._artifactDir, reference)
    const resolvedDir = path.resolve(this._artifactDir)
    if (!filePath.startsWith(resolvedDir)) {
      throw new Error(`Reference not found: ${reference}`)
    }

    const filename = path.basename(filePath)

    try {
      const content = await fs.readFile(filePath)
      const contentType = this._contentTypes[filename] ?? 'application/octet-stream'
      return { content: new Uint8Array(content), contentType }
    } catch {
      throw new Error(`Reference not found: ${reference}`)
    }
  }
}

/**
 * S3-based storage backend.
 *
 * Stores offloaded content as S3 objects. Content type is preserved as S3 object metadata.
 * References are `s3://` URIs for direct access via AWS CLI or SDK.
 *
 * @param bucket - S3 bucket name
 * @param options - Optional configuration (prefix, region, pre-configured S3Client)
 */
export class S3Storage implements Storage {
  private readonly _bucket: string
  private readonly _prefix: string
  private _client: import('@aws-sdk/client-s3').S3Client | undefined
  private readonly _region: string
  private _counter = 0

  constructor(
    bucket: string,
    options?: { prefix?: string; region?: string; s3Client?: import('@aws-sdk/client-s3').S3Client }
  ) {
    this._bucket = bucket
    this._prefix = options?.prefix ? options.prefix.replace(/\/+$/, '') + '/' : ''
    this._client = options?.s3Client
    this._region = options?.region ?? 'us-east-1'
  }

  private async _getClient(): Promise<import('@aws-sdk/client-s3').S3Client> {
    if (this._client) return this._client
    const { S3Client } = await import('@aws-sdk/client-s3')
    this._client = new S3Client({ region: this._region })
    return this._client
  }

  /** {@inheritdoc} */
  async store(key: string, content: Uint8Array, contentType: string = 'text/plain'): Promise<string> {
    const client = await this._getClient()
    const { PutObjectCommand } = await import('@aws-sdk/client-s3')

    const sanitizedKey = sanitizeId(key)
    const timestampMs = Date.now()
    this._counter++
    const s3Key = `${this._prefix}${timestampMs}_${this._counter}_${sanitizedKey}`

    await client.send(
      new PutObjectCommand({
        Bucket: this._bucket,
        Key: s3Key,
        Body: content,
        ContentType: contentType,
      })
    )

    return `s3://${this._bucket}/${s3Key}`
  }

  /** {@inheritdoc} */
  async retrieve(reference: string): Promise<{ content: Uint8Array; contentType: string }> {
    const client = await this._getClient()
    const { GetObjectCommand } = await import('@aws-sdk/client-s3')

    // Accept both s3:// URIs and raw keys
    let s3Key = reference
    const uriMatch = reference.match(/^s3:\/\/([^/]+)\/(.+)$/)
    if (uriMatch?.[1] && uriMatch[2]) {
      if (uriMatch[1] !== this._bucket) {
        throw new Error(`Reference not found: ${reference} (bucket mismatch)`)
      }
      s3Key = uriMatch[2]
    }

    try {
      const response = await client.send(
        new GetObjectCommand({
          Bucket: this._bucket,
          Key: s3Key,
        })
      )
      const body = await response.Body?.transformToByteArray()
      if (!body) throw new Error(`Reference not found: ${reference}`)
      const contentType = response.ContentType ?? 'application/octet-stream'
      return { content: new Uint8Array(body), contentType }
    } catch (error: unknown) {
      if (error instanceof Error && error.name === 'NoSuchKey') {
        throw new Error(`Reference not found: ${reference}`)
      }
      throw error
    }
  }
}
